docs/cpu-metrics-in-fleet.md

```diff
@@ -0,0 +1,374 @@
+      Elasticsearch query.
+   5. this per-10-second CPU utilization (in milliseconds) is divided by `10 * 1000` (= 10 seconds expressed as milliseconds),
+      to get the _percentage_ CPU utilization over 10 elapsed seconds. This is the `cpu` aggregation in the Elasticsearch query.
+   6. So now we end up with five CPU utilization (in %) values, one for each minute.
+   7. These are averaged to arrive at the CPU utilization (in %) over five minutes. This is the `avg_cpu` aggregation in
+      the Elasticsearch query.
+2. The 5-minute average CPU utilization (in %) for each of the processes is summed up to arrive at a total CPU utilization
+   (in %) for the Agent process and all Beat processes managed by Agent. This is the `sum_cpu` aggregation in the Elasticsearch
+   query. The result is the total CPU utilization (in %) by Agent and it's child Beat processes across all available cores
+   (as opposed to being normalized per-core). For example, if a machine has 8 cores, the resulting value will be in the range of
+   (0%, 800%).
+
+## Observations
+
+* The CPU utilization (in %) of Agent and each of the Beat processes can vary wildly. Generally speaking, the Agent process
+  itself does not utilize much CPU. And each Beat process may utilize CPU depending on the type of computations it is performing
+  on the data.
+
+* Also, CPU utilization is rarely constant. If the output of `top` or `htop` for Agent and Beat processes is observed over
+  time, the CPU utilization % shown varies for each process over time.
+
+* To relate the output seen in `top` or `htop` for Agent and Beat processes with the single value shown in the Fleet UI,
+  one must observe the values in the `top` / `htop` output over five minutes, arrive at an average CPU utilization (in %)
+  value for each process, and sum it up.  The resulting value will roughly be equal to the single value shown for that Agent
+  in the Agent Listing page in the Fleet UI.
+
+## Suggested improvements
+
+* We should document the observations above in an appropriate location and perhaps link to this documentation from the "i"
+  icon in CPU column in the Agent Listing page in the Fleet UI.
+
+* Rather than showing a single value for every Agent in the Agent Listing page in the Fleet UI, we should consider showing
+  the constituent values for Agent and each of the Beats processes managed by the Agent. Concretely, this would mean
+  getting rid of the `sum_cpu` aggregation in the Elasticsearch query. The challenge, of course, would be in presenting
+  these values in a user-friendly manner, especially in cases where an Agent is managing several Beats. Also, by showing
+  the values for the Beat processes (along with the Agent process), we would be exposing internals of Agent to users, which
+  may be confusing for some users.
+
+* We should enhance collection and aggregation to include CPU utilization for Beats, which are Agent components managed by
```

Comment:
Can you elaborate on this? You mean provide the CPU metrics in the `stats` metrics, instead of requiring fleet to perform its own aggregation query?

Reply:
Sorry, I don't think I wrote this in the clearest way.  I've changed the language now but what I wanted to convey was that we should also include and present CPU metrics about service-runtime components (e.g. Endpoint), not just command-runtime components (e.g. Beats).  And I deliberately didn't want to get into exactly _how_ we should do that.

Reply:
I think we account for endpoint CPU now as well, which will even further complicate users attempt to correlate what is shown in Fleet to what is in top.

Top will never account for endpoint resource usage because it is in a different process and cgroup.

Reply:
Hmmm, I'm not seeing Endpoint CPU being recorded today.

I just added the Elastic Defend integration to my policy and then, a few minutes later, queried the `metrics-elastic_agent*` indices with the following query:

```
POST metrics-elastic_agent*/_search?filter_path=hits.hits._source
{
  "sort": [
    {
      "@timestamp": {
        "order": "desc"
      }
    }
  ],
  "_source": [
    "@timestamp", 
    "elastic_agent.process",
    "component.id"
  ],
  "collapse": {
    "field": "component.id"
  },
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "elastic_agent.id": [
              "c406741d-7673-447f-b7e0-43417eeb10d6"
            ]
          }
        }
      ]
    }
  }
}
```

Results:

```json
{
  "hits": {
    "hits": [
      {
        "_source": {
          "@timestamp": "2024-01-10T22:03:45.069Z",
          "component": {
            "id": "log-default"
          },
          "elastic_agent": {
            "process": "filebeat"
          }
        }
      },
      {
        "_source": {
          "@timestamp": "2024-01-10T22:03:45.069Z",
          "component": {
            "id": "elastic-agent"
          },
          "elastic_agent": {
            "process": "elastic-agent"
          }
        }
      }
    ]
  }
}```